### ISSUE TITLE
How to create a fake SocketModeClient for testing? (send fake MessageEvent socket message)

### ISSUE THREAD

**Issue by chrisyarbrough** [soxtoby/SlackNet#220]
Is there a way I can override the SlackSocketModeClient with a fake instance for manual testing? I would want to build a CLI helper that I can type messages to and they would be sent to the rest of my bot code in place of using the real Slack client app.

I'm currently following the examples closely and started with this:

```
using Autofac;
using SlackNet.Autofac;
using SlackNet.SocketMode;

class MyProgram
{
	public async void Start()
	{
		var containerBuilder = new ContainerBuilder();
		containerBuilder.AddSlackNet(c => c
			.RegisterEventHandler<MessageEvent, PingDemo>()
			.UseSocketModeClient<FakeSlackSocketModeClient>()
		);
		await using IContainer container = containerBuilder.Build();
		ISlackSocketModeClient? client = container.SlackServices().GetSocketModeClient();
		await client.Connect();
		Console.ReadKey();
	}
}

class FakeSlackSocketModeClient : ISlackSocketModeClient
{
	public Task Connect(
		SocketModeConnectionOptions? connectionOptions = null,
		CancellationToken cancellationToken = new CancellationToken())
	{
		Connected = true;
		return Task.CompletedTask;
	}

	public void Dispose()
	{
	}

	public void Disconnect()
	{
		Connected = false;
	}

	public bool Connected { get; private set; }
}

class PingDemo : IEventHandler<MessageEvent>
{
	public Task Handle(MessageEvent slackEvent)
	{
		Console.WriteLine("Received message: " + slackEvent.Text);
		return Task.CompletedTask;
	}
}
```

To me it looks like SlackNet provides all sorts of DI insertion points, but I'm lost when it comes to the part where I emulate a fake socket message.

I believe I want to create a fake payload like this:

```
var messageEvent = new MessageEvent()
{
	User = "U12345",
	Channel = "C12345",
	Text = "This is fake message."
};

var socketMessage = new EventEnvelope()
{
	EnvelopeId = Guid.NewGuid().ToString(),
	Payload = new EventCallback()
	{
		Type = "message",
		Event = messageEvent
	}
};
```

But now, how do I send this as if the Slack client would have sent it for processing into the rest of SlackNet? My hope would be, that I can add the fake socket message somehow, but still keep the rest of the event handler registrations etc in place, so that I can test my handlers without having to use the real Slack client.

**Comment by soxtoby** [comment]
Hi @chrisyarbrough, interesting problem!

Assuming you can't just resolve your event handler and pass the events in directly, I think you've got a couple options.

One is to implement the core of what `SlackSocketModeClient` does when messages come in. You won't need a fake `SlackSocketModeClient` for this - just your own class that takes in an event and passes it to the appropriate handler.

```c#
var requestListeners = container.SlackServices().GetRequestListeners();
var handlerFactory = container.SlackServices().GetHandlerFactory();

// For each event
var request = new SlackRequestContext();
await using (request.BeginRequest(requestListeners))
    await handlerFactory.CreateEventHandler(request).Handle(new EventCallback { Event = new MessageEvent { Text = "test" } });
```

Another is to use a real `SlackSocketModeClient` with a fake `CoreSocketModeClient`.
```c#
class FakeSocketModeClient : ICoreSocketModeClient
{
    public void Receive(SocketMessage message) => _messages.OnNext(message);
    
    private readonly Subject<SocketMessage> _messages = new();
    public IObservable<SocketMessage> Messages => _messages;

    public Task Connect(SocketModeConnectionOptions? connectionOptions = null, CancellationToken cancellationToken = default) => Task.CompletedTask;
    public void Disconnect() { }
    public bool Connected => true;
    public void Send(int socketId, Acknowledgement acknowledgement) { }
    public IObservable<RawSocketMessage> RawSocketMessages { get; } // Can leave this as null
    public void Dispose() { }
}
```
Register it with:
```c#
var socketModeClient = new FakeSocketModeClient();
containerBuilder.AddSlackNet(c => c
    .UseSocketModeClient((ISlackServiceProvider sp) => new SlackSocketModeClient(
        socketModeClient, 
        sp.GetJsonSettings(), 
        sp.GetRequestListeners(), 
        sp.GetHandlerFactory(), 
        sp.GetLogger()))
);
```
and trigger it with
```c#
socketModeClient.Receive(new EventEnvelope { Payload = new EventCallback { Event = new MessageEvent { Text = "test" } } });
```

Hope that helps.

**Comment by chrisyarbrough** [comment]
Yea, it's a fringe case, we want our "Slack" to be available offline, without access to the workspace. Creating unit-testing code for each handler, etc. is definitely a good way to go anyway, but I thought it could help us with prototyping if we had this fake implementation working in a way, that doesn't touch much of the rest of the app.

Thanks for the guidance, that already helps a bunch! :) 

I kind of like the idea of just manually sending the events to the handlers, that's far less code that I thought, and probably already does what we need!

I also setup a FakeCoreSocketModeClient and found it helpful. The only caveat is that, we also stepped into the rabbit hole of having to fake the SlackAPI and reroute the bot response to the Console. Not so sure if my plan is as "lean" as I originally thought.

In any case, thanks for the support!

For anyone who might want to do a similar thing, here's what I have, while still trying out the earlier suggestions:
To also fake the ApiClient, it seems I needed to use Reflection, or I don't know how to correctly fit into the DI API:
```
public static void Configure(ISlackServiceProvider slackServices)
{
  object baseProvider = ReflectionUtil.GetPrivateFieldValue(slackServices, "_baseProvider")!;
  
  // This part can be replaced by .UseSocketModeClient (see soxtoby above)
  ReflectionUtil.SetPrivateFieldValue(baseProvider, "_socketModeClient",
	  new Lazy<ISlackSocketModeClient>(() => CreateSocketModeClient(slackServices)));
  
  ReflectionUtil.SetPrivateFieldValue(baseProvider, "_apiClient",
	  new Lazy<ISlackApiClient>(CreateMockSlackApiClient));
}
```

I then used the moq framework:
```
private static ISlackApiClient CreateMockSlackApiClient()
{
  var mockSlackApiClient = new Mock<ISlackApiClient>();
  
  mockSlackApiClient.Setup(client => client.Chat).Returns(FakeChatClient.MockChatApi());
  mockSlackApiClient.Setup(client => client.Reactions).Returns(FakeReactionClient.MockReactionsApi());
  mockSlackApiClient.Setup(client => client.Views).Returns(Mock.Of<IViewsApi>());
  mockSlackApiClient.Setup(client => client.Conversations).Returns(Mock.Of<IConversationsApi>());
  
  return mockSlackApiClient.Object;
}
```

Which allows us to, instead of posting to the real Slack, just reroute bot responses to the console:
```
public static IChatApi MockChatApi()
{
  var mockChatApi = new Mock<IChatApi>();
  
  mockChatApi
	  .Setup(chat => chat.PostEphemeral(It.IsAny<string>(), It.IsAny<Message>(), It.IsAny<CancellationToken>()))
	  .Callback<string, Message, CancellationToken>((_, message, _) => { Console.WriteLine(message.Text); })
	  .Returns(Task.FromResult(new PostEphemeralResponse()));
  
  mockChatApi
	  .Setup(chat => chat.PostMessage(It.IsAny<Message>(), It.IsAny<CancellationToken>()))
	  .Callback<Message, CancellationToken>((message, _) => { Console.WriteLine(message.Text); })
	  .Returns(Task.FromResult(new PostMessageResponse()));
  
  return mockChatApi.Object;
}
```

So, basically this is what you would also do if you wanted to unit-test your event handlers, but in my case, the whole thing works in the runtime, which allows both runtime and tests to use a single composition root and not touch too much within the production code of the app afterwards.

But most of the time it will probably make more sense to just use regular unit-testing with mocks in the test project, instead of integration testing the whole app as one piece.
